### DISCUSSION TITLE
Unifying structs and interfaces

### DISCUSSION THREAD

**Discussion by MarkMcCulloh** [winglang/wing#5102] [General]
Started a discussion in https://github.com/winglang/wing/issues/4961

Reposting my comment below:

In addition to this change, I think we should evaluate the purpose of structs. In terms of type expressiveness, this change makes structs a subset of interfaces. structs no longer serve a purpose that can't be solved by the compiler itself. Why not get rid of structs and support interfaces everywhere structs are used i.e.:

```
// maybe we stop calling them interfaces and call them `type` or `shape` instead
interface Whatever {
  field: num;
  func(): void;
}

let x = Whatever {
  field: 2,
  func() {}, // or `func: () => {},` I guess
}
```

If the concern is "structs are special because they provide a guarantee of no functions/behavior", I would say:
1. We're statically typed, the compiler knows if interfaces do/don't have functions
2. This isn't even true to begin with, as structs can have class instances as fields

--- 

and also a response from @eladb :

there's a fundamental difference between behavioral interfaces and data structures in my mind. i don't think structs and interfaces are the same thing. for example, Struct.fromJson doesn't make any sense on interfaces.

---

My response to the response:

> there's a fundamental difference between behavioral interfaces and data structures in my mind. i don't think structs and interfaces are the same thing.

The key part of this is that the difference is in the mind, they don't actually need to do anything different. Functionally structs and interfaces "can" be the same thing. That fact that you decided that interfaces are behavioral is arbitrary once we allow properties. It means I can create an interface with only properties. Is that behavioral? Sure, but then structs are also behavioral.

> Struct.fromJson doesn't make any sense on interfaces.

If you're saying that because interfaces allow functions, then I would argue it doesn't make sense on structs either because structs can still contain functions (via classes). If you have an interface with only properties, sure why not allow fromJson. Either way, this can be statically determined.

I think we're on the same page with interfaces (no need to repeat that they are behavioral or pointing to anything "fundamental", I understand), my concern is with structs. They aren't limited enough. Structs should not allow classes.
If we don't restrict them further, I would argue there is too much overlap between structs and interfaces and we might as well get rid of structs.

**Comment by eladb** [comment]
Yes, structs should be more limited and we will add these constraints in the future, but I don't think we should get rid of them because we can't fully express these concerns yet.

In my mind the difference between interfaces and structs is that interfaces represent objects that **must** be passed by reference while structs represent objects that **can** be passed by-value (by bitwise copying all the data).

Currently, we can't really express this constraint in the language but ideally structs should only reference *immutable* objects. Currently, we don't really have immutability modeled for user-defined types only for the built-in collection types, but once we introduce that, then I would add that constraint to structs.

**Comment by Chriscbr** [comment]
The idea of combining structs/interfaces is starting to grow on me.

Here's a (slightly modified) proposal that uses a new word, "shape", to try to put aside any preconceived assumptions you have from words like "interface" or "struct". But the name isn't essential to the design.

----

Instead of interfaces or structs, Winglang supports shapes. A shape is a contract that allows you specify that a value has a set of publicly available members. A shape is defined like so:

```js
shape AwsInstance {
  publicIp: str?;
  start: inflight(): void;
  stop: inflight(): void;
}
```

> This syntax emphasizes that the members `start` and `stop` don't have any particular phase associated with them, only their values do.

A shape does not have a phase associated with it. But function members defined on the shape may have phases, like any other function annotations.

Shapes are implemented by classes. A shape can by satisfied by an instance of a preflight class, and inflight class, or an anonymous class object:

```js
class A impl AwsInstance {
  publicIp: str?;
  inflight start(): void { ... }
  inflight stop(): void { ... }
}

// preflight class object
let a: AwsInstance = new A();

inflight class B impl AwsInstance {
  ...
}

// inflight class object
let b: AwsInstance = new B();

// "anonymous class object"
let c = AwsInstance {
  publicIp: nil,
  start: inflight () => { ... },
  stop: inflight () => { ... },
};

// equivalent to...
/*
class $AwsInstance1 impl AwsInstance {
  publicIp: str = nil;
  inflight start() { ... }
  inflight stop() { ... }
}
let c = new $AwsInstance1();
*/
```

Shapes can extend other shapes. Shape members can only add members - a shape cannot override members of shapes it extends:

```js
shape Foo {
  foo: str;
}

shape Bar extends Foo {
  foo: str;
}
```

If a shape has preflight methods or if it extends the "construct" shape, then it cannot be implemented by an inflight class:

```js
shape IBucket extends construct.Construct {
  addObject: (): void; // implicitly preflight
}

inflight class B impl IBucket { // error: members "addObject" of the shape IBucket prevent it being implemented by an inflight class
  ...
}

// alternate syntax:
construct shape IBucket {
  ...
}
```

If a function parameter has a shape type, it can accept any value that implements that shape:

```js
let restartInstance = inflight (inst: AwsInstance) => {
  inst.stop();
  inst.start();
};
```

Shapes can serve additional purposes. One use for shapes are named parameters in method signatures or class constructors:

```js
shape FooProps {
  prop1: bool;
  prop2: inflight (): void;
}

class Foo {
  new(props: FooProps) {}
}

new Foo(prop2: inflight () => { ... }, prop1: true);
```

If the `json` modifier is added to a shape declaration, then `.fromJson()` and `.toJson()` methods will be automatically available on the shape type:

```js
json shape Order {
  id: str;
  items: Array<str>;
}

Order.fromJson(...);
```

The `json` modifier can only be added to shapes whose member types are serializable over the wire.

**Reply by MarkMcCulloh** [reply]
I love it, this is exactly what I had in mind

**Reply by eladb** [reply]
I am warming up to this, but something is still not sitting well in my head.

How do I define "shaped data"? If the only way to instantiate an object that implements a shape is through a class, then what's the syntax?

That's how it's done today:

```js
let order = Order { ... };
```

Is this still relevant? Is this basically a sugar for:

```js
let order = class impl Order { };
```

**Reply by MarkMcCulloh** [reply]
I don't actually think the "anonymous class object" is necessary for implementation.

The current syntax we have for something like `let order = Order { ... };` or `fn(field: ..., method: ...)` works perfectly fine for the shape concept without creating classes.

The only thing that matters in the end is if the object matches the contract, whether the object is a class instance or a plain object isn't relevant.

**Reply by Chriscbr** [reply]
It would basically be a sugar, yes. Having a terse way to create shape objects (without having to write boilerplate code for an entire class) would still be useful in my mind.

One piece of ambiguity I thought of is how equality is defined on shapes. (today, equality on instances of classes is done by reference, and equality on structs has been, at least initially, defined as comparing all of the fields - see https://www.winglang.io/docs/language-reference#617-structs). But I think for now we can start by just saying all shape objects are compared by reference (i.e. if they refer to the same JS object under the hood) and add some kind of custom type equality mechanism later. Or perhaps we could say that `json shape Foo {}` / `data shape Foo {}` types have a different equality mechanism.

**Reply by eladb** [reply]
The by ref/by value issue smells like it could be a major issue.

**Reply by MarkMcCulloh** [reply]
We already have this smell with structs since they allow classes as fields, so it's only partially value equality. If we do eventually stop structs from allowing class fields, we're nonetheless going to add fields to interfaces. Would that mean interfaces are partially checked by value too? What about an interface made up only fields?

IMO referential equality is a perfectly fine default for all non-builtins as long as value equality is also easy to do with a well-explained api. Like through a global func or util module. Value equality is a bad default because it's slow and adds ambiguity.

> Or perhaps we could say that `json shape Foo {}` / `data shape Foo {}` types have a different equality mechanism.

If want to add an automatic by-value data structure I think this is ok.

This is highlighting why I made this discussion in the first place. 
`data shape Foo {}` is a "struct/record/pass-by-value-thing" and `shape Foo {}` is an "interface/type" but they're both still just contracts. The difference is so minimal that it can be reasonably expressed in a single modifier, where `data` enforces "Foo's equality is by value and its properties can be serialized".

There's no reason you can't do `let x = Foo { ... }` or `class X impl Foo {}` regardless of that constraint being present or not.

(As a side note, I feel like we're misusing "by-value" and "by-reference" here. I'm not convinced wing will _ever_ have memory-level semantics like that.)

**Reply by Chriscbr** [reply]
Some other open questions:

- Should there be some kind of rule that `data shape` types can't be implemented by classes? I'm not sure how auto-generated toJson/fromJson static methods would work since your class may have private fields.
- Our language doesn't have computed field support yet, but it's on our roadmap. How would that behave with shapes? For example, could this be allowed? Or does it have to be opt-in per field?

```
shape Foo {
  n: num;
}
let x: Foo = ...
x.n = 123;
```

**Reply by eladb** [reply]
If we merge interfaces and structs we won't be able to get away with `impl Interface`, no?

BTW, an alternative name for this would be `protocol`

**Reply by MarkMcCulloh** [reply]
> If we merge interfaces and structs we won't be able to get away with impl Interface, no?

This looks fine to me

```ts
shape Foo {
  n: num;
}

class Stuff impl Foo {
  pub n: num = 2;
}
```

In the end this `shape` thing is basically just what typescript interfaces are, with the addition that we get to use some our cool sugary `struct` syntax with it.

**Reply by eladb** [reply]
How about `protocol` and `wire protocol` (`wire` instead of `json`)

```js
protocol IBucket {

}

wire protocol Employee {
  name: str;
  last: str;
}
```

Then `wire` implies ser/deser methods and capabilities.

**Reply by MarkMcCulloh** [reply]
I'm fine to get swifty with `protocol`. 
`wire` feels a bit off, it implies that networking must be involved in some way. I'd prefer `data` or `record`

**Reply by eladb** [reply]
+1 for `data protocol`. I really like it

**Reply by yoav-steinberg** [reply]
Can classes or non-data protocols be members of a data protocol?
If not I feel like "data protocols" are what we originally intended for structs. But in any case we should be clear about this.
Also can a data protocol contain functions?

**Reply by eladb** [reply]
I really like the idea of introducing `protocol` and `data protocol` and getting rid of `struct`s and `interface`s. I think that's a better model.

**Reply by Chriscbr** [reply]
Another open question -- we recently introduced inflight interfaces as a complement to [implicitly-]preflight interfaces -- how does that factor in here?

**Reply by eladb** [reply]
In a sense an inflight interface is merely sugar for `interface extends std.IResource`, correct?

So either we just continue to support this for protocols (`inflight protocol`) or:

```js
protocol Boo extends std.IResource
```

**Reply by thoroc** [reply]
> A shape is a contract

Sorry to wander in a bit late in the conversation, but since you are talking about `contract` and using the keyword `shape`, would it be more sensical to just use the keyword `contract` instead?

I appreciate if you've made the changes and your mind up and this suggestion is only adding noise.

**Reply by Chriscbr** [reply]
@thoroc No worries - naming suggestions are welcome. `contract` is also reasonable. I think to my ears one benefit of `shape` is it connotes the structural typing behavior (if two shapes have the same contents, then they're interchangeable). But I'm also liking the sound of `protocol`
